source/common/http/filter/cors_filter.cc

```diff
@@ -0,0 +1,152 @@
+#include "common/http/filter/cors_filter.h"
+
+#include "envoy/http/codes.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+
+namespace Envoy {
+namespace Http {
+
+CorsFilter::CorsFilter(CorsFilterConfigConstSharedPtr config)
+    : config_(config), is_cors_request_(false) {}
+
+CorsFilter::~CorsFilter() {}
+
+FilterHeadersStatus CorsFilter::decodeHeaders(HeaderMap& headers, bool) {
+  origin_ = headers.Origin();
+  if (origin_ == nullptr || origin_->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  initialize();
+
+  if (!enabled()) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (!isOriginAllowed(origin_->value())) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  is_cors_request_ = true;
+
+  auto method = headers.Method();
+  if (method == nullptr || method->value().c_str() != Http::Headers::get().MethodValues.Options) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  auto requestMethod = headers.AccessControlRequestMethod();
+  if (requestMethod == nullptr || requestMethod->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  HeaderMapPtr response_headers{
+      new HeaderMapImpl{{Headers::get().Status, std::to_string(enumToInt(Http::Code::OK))}}};
+
+  response_headers->insertAccessControlAllowOrigin().value(*origin_);
+
+  if (allowCredentials()) {
+    response_headers->insertAccessControlAllowCredentials().value(
+        Http::Headers::get().CORSValues.True);
+  }
+
+  if (!allowMethods().empty()) {
+    response_headers->insertAccessControlAllowMethods().value(allowMethods());
+  }
+
+  if (!allowHeaders().empty()) {
+    response_headers->insertAccessControlAllowHeaders().value(allowHeaders());
+  }
+
+  if (!exposeHeaders().empty()) {
+    response_headers->insertAccessControlExposeHeaders().value(exposeHeaders());
+  }
+
+  if (!maxAge().empty()) {
+    response_headers->insertAccessControlMaxAge().value(maxAge());
+  }
+
+  decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterHeadersStatus CorsFilter::encodeHeaders(HeaderMap& headers, bool) {
+  if (!is_cors_request_) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  headers.insertAccessControlAllowOrigin().value(*origin_);
+  if (allowCredentials()) {
+    headers.insertAccessControlAllowCredentials().value(Http::Headers::get().CORSValues.True);
+  }
+
+  return FilterHeadersStatus::Continue;
+}
+
+void CorsFilter::setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+};
+
+void CorsFilter::initialize() {
+  route_cors_policy_ = decoder_callbacks_->route()->routeEntry()->corsPolicy();
```

Comment:
Both of these calls aren't always safe. The route() might be empty. It might also be a redirect. You need to handle both of these cases.

Reply:
I will skip the filter in case it's a redirect, since the final host should probably handle that. Is that ok?

Reply:
Yes, that's what I would do. (Skip of no route or redirect)